### ISSUE TITLE
Don't use pipeline::Value where it isn't warranted

### ISSUE THREAD

**Issue by larrylindsey** [catsop/catsop#29]
Use boost::shared_ptr instead

**Comment by larrylindsey** [comment]
A question for @funkey :

In the speed-up branch, line 284 in SegmentGuarantor looks like:

```
_segmentStore->associate(segments, *block);
```

Where segments is of type Segments, not pipeline::Value<Segments>, and dereferencing block gets you a Block.

The associated call in SegmentStore.h:34 looks like:

```
virtual void associate(pipeline::Value<Segments> segments,
                           pipeline::Value<Block> block) = 0;
```

How is this conversion magic done, or am I missing something? Also, why abandon passing shared_ptrs here?

**Comment by funkey** [comment]
The conversion is handled by the

```
Value(const T& value) {...}
```

constructor of Value<T>, which creates a new pipeline::Value from an existing object.

Passing shared_ptr<T>s would be better here, to avoid the pipeline::Value<T> overhead. In this particular case, I would even change the interface of associate to accept references:

```
virtual void associate(const Segments& segments, const Block& block) = 0;
```

**Comment by funkey** [comment]
By the way, here is why I prefer references in class interfaces:
1. References are more general.
2. Method signatures are easier to read.
3. Copy overhead is avoided.

In detail:
1. Any pointer and any object can be cast into a reference. If a reference is 
   all the callee needs, all those cases are covered already. If the interface 
   requires a shared pointer (or a pipeline::Value), it might at least be 
   inefficient to use it for objects that are stack allocated on the caller side.
2. It is clear from the signature what the method does. When we pass an object 
   by reference, it should be clear that our object is only visiting the other 
   scope, but ownership is exclusive to us and we manage the lifetime of the 
   object. The const keyword gives additional hints which of the passed objects are 
   read-only.
3. We avoid calling the copy constructor of shared pointers (or 
   pipeline::Values). The resulting code might be easier for the compiler to 
   optimize. In addition, consequent use of the const keyword gives optimization 
   hints to the compiler.

In summary: Pass a shared pointer if your intention is to share ownership of the 
object between the caller and the callee. For all other cases, use references 
and be consequent in the use of the const keyword.

**Comment by larrylindsey** [comment]
TIL: There are such things as converting constructors in C++.

The preference for references makes sense to me. My concern is that we'll come across a situation in which an object receiving a call needs a permanent reference to an argument that has been passed as reference earlier in the call-chain. I'll address those situations as they arise.

Thanks Jan
